### PR TITLE
psa_switch support recirculate

### PR DESCRIPTION
**Pull request by derekso1**
Changes:

- Added support for packet recirculation in psa_switch
- accompanying STF test is currently in [this PR](https://github.com/p4lang/p4c/pull/1857)

### CODE REVIEW

File: targets/psa_switch/psa_switch.cpp [hunk 1]
```diff
@@ -396,6 +396,20 @@ PsaSwitch::egress_thread(size_t worker_id) {
     Pipeline *egress_mau = this->get_pipeline("egress");
     egress_mau->apply(packet.get());
     deparser->deparse(packet.get());
+    
+    if (port == PSA_PORT_RECIRCULATE) {
```
Comment (derekso1) [inline targets/psa_switch/psa_switch.cpp]: Currently this does a packet copy whenever we recirculate a packet. This is very similar to how simple_switch handles recirculation. Seems like its a possibility to remove the packet copying? Thoughts?

File: targets/psa_switch/psa_switch.cpp [hunk 2]
```diff
@@ -396,6 +396,20 @@ PsaSwitch::egress_thread(size_t worker_id) {
...
+      BMLOG_DEBUG_PKT(*packet, "Recirculating packet");
+      std::unique_ptr<Packet> packet_copy = packet->clone_no_phv_ptr();
+      PHV *phv_copy = packet_copy->get_phv();
+      phv_copy->reset_metadata();
+      phv_copy->get_field("psa_ingress_input_metadata.packet_path")
```
Comment (derekso1) [inline targets/psa_switch/psa_switch.cpp]: As mentioned in a previous PR, we should be using the PACKET_PATH enum from JSON. Compiler support is currently pending approval on p4c PR.

Comment (jafingerhut) [reply targets/psa_switch/psa_switch.cpp]: I think there are 2 different packet_path metadata fields that must be assigned here, the one you have already, plus the one in `psa_ingress_parser_input_metadata`

Comment (jafingerhut) [reply targets/psa_switch/psa_switch.cpp]: The PSA spec also says that recirculated packets should have both of these fields assigned the value of PSA_RECIRCULATE_PORT for recirculated packets.  That should be done somewhere near this line of code, too:

+ psa_ingress_parser_input_metadata.ingress_port
+ psa_ingress_input_metadata..ingress_port

Comment (jafingerhut) [reply targets/psa_switch/psa_switch.cpp]: Another field: psa_ingress_input_metadata.ingress_timestamp should have the current time stored into it when a packet is recirculated, according to the PSA spec.

File: targets/psa_switch/psa_switch.cpp (hunk 1 above)
Comment (jafingerhut) [reply targets/psa_switch/psa_switch.cpp]: This comment isn't helpful, other than to say I do not know the answer here.  I am hoping @antoninbas might have time to look at this question and say what he knows of it?

Comment (antoninbas) [reply targets/psa_switch/psa_switch.cpp]: By the way you a `reset_metadata` but aren't you supposed to preserve the recirculation metadata?

Comment (jafingerhut) [reply targets/psa_switch/psa_switch.cpp]: I think the idea is to get recirculate working first, without preserving metadata, and to save for later whatever changes are required to get recirculate while preserving metadata working.

Comment (derekso1) [reply targets/psa_switch/psa_switch.cpp]: Yes that is the plan.

**Comment by derekso1** [comment]
I removed packet cloning in recirculation and updated the psa metadata (apologies slipped my mind). Also updated the ingress timestamp in receive_ since that was not done previously.

**Pull request merged**